File: src/Sonic4Episode2.Core/Engine/DeathWait.cs
namespace Sonic4Episode2.Core.Engine;

public enum DeathWaitAction
{
    None,
    Restart,
    GameOver,
}

public readonly record struct DeathWaitState(float Elapsed, int RemainingLives);
public readonly record struct DeathWaitResult(DeathWaitState State, DeathWaitAction Action);

public static class DeathWait
{
    public const float Duration = 120f;

    public static DeathWaitResult Tick(DeathWaitState state, bool dead, bool suppressed,
                                       float timeStep = 1f)
    {
        if (!float.IsFinite(state.Elapsed) || state.Elapsed < 0 ||
            !float.IsFinite(timeStep) || timeStep < 0 ||
            state.RemainingLives < -1 || (state.RemainingLives == -1 && state.Elapsed < Duration))
            throw new ArgumentOutOfRangeException(nameof(state));
        if (!dead || suppressed || state.Elapsed >= Duration)
            return new(state, DeathWaitAction.None);

        float elapsed = state.Elapsed + timeStep;
        if (!float.IsFinite(elapsed)) throw new ArgumentOutOfRangeException(nameof(timeStep));
        if (elapsed < Duration)
            return new(new(elapsed, state.RemainingLives), DeathWaitAction.None);

        int lives = state.RemainingLives - 1;
        return new(new(elapsed, lives), lives < 0 ? DeathWaitAction.GameOver : DeathWaitAction.Restart);
    }
}

File: src/Sonic4Episode2.Native/src/death_wait.cpp
#include "death_wait.h"

#include <cmath>
#include <stdexcept>

namespace {

void validate_state(const DeathWaitState& state, float time_step) {
    if (!std::isfinite(state.elapsed) || state.elapsed < 0.0f ||
        !std::isfinite(time_step) || time_step < 0.0f ||
        state.remaining_lives < -1 ||
        (state.remaining_lives == -1 && state.elapsed < kDeathWaitDuration)) {
        throw std::invalid_argument("invalid death wait state");
    }
}

}

DeathWaitAction tick_death_wait(
    DeathWaitState& state,
    bool dead,
    bool suppressed,
    float time_step) {
    validate_state(state, time_step);

    if (!dead || suppressed || state.elapsed >= kDeathWaitDuration) {
        return DeathWaitAction::None;
    }

    const float elapsed = state.elapsed + time_step;
    if (!std::isfinite(elapsed)) {
        throw std::invalid_argument("death wait elapsed overflow");
    }

    state.elapsed = elapsed;
    if (state.elapsed < kDeathWaitDuration) {
        return DeathWaitAction::None;
    }

    --state.remaining_lives;
    return state.remaining_lives >= 0 ? DeathWaitAction::Restart : DeathWaitAction::GameOver;
}
